### PR TITLE
BSP part of IOT2050-SM

### PR DESCRIPTION
**Pull request by BaochengSu**
This is the BSP part of the IOT2050-SM: u-boot and kernel.

### CODE REVIEW

File: recipes-bsp/u-boot/files/0008-board-siemens-iot2050-Pass-DDR-size-from-FSBL.patch [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+--- a/doc/board/siemens/iot2050.rst
++++ b/doc/board/siemens/iot2050.rst
+@@ -29,6 +29,11 @@ The following binaries from that source need to be present in the build folder:
+  - seboot_pg1.bin
+  - seboot_pg2.bin
+ 
++Starting from SE-Boot D/V01.04.0x.0x, the DDR size is passed from the SE-Boot to
++the u-boot, this is not compatible with the old version u-boot (<=2023.10)
++anymore. Make sure to use the latest u-boot version, or version contains the
++commit "board: siemens: iot2050: Pass DDR size from FSBL".
```
Comment (jan-kiszka) [inline recipes-bsp/u-boot/files/0008-board-siemens-iot2050-Pass-DDR-size-from-FSBL.patch:101]: "Note that SE-Boot D/V01.04.01.02 or greater is required, otherwise the DDR size
will not be picked up correctly by U-Boot."

I makes no sense to ask for a U-Boot version greater than X in the U-Boot source tree.

Also, we do not release Vxx

File: recipes-bsp/u-boot/files/0009-board-siemens-iot2050-Generalize-the-fdt-fixup.patch [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+         E-key: PCIe, USB 2.0
+ "2"  -  B-key: USB 3.0,
+         E-key: PCIe, USB 2.0
++
++Device tree overlay fixup
++-------------------------
++
++In rare cases, you may need to add your own device tree overlay, please take use
++of the **do_overlay_prepare** and **variants_fdt_fixup** from
++**board/siemens/iot2050/board.c** to add overlay fixup.
```
Comment (jan-kiszka) [inline recipes-bsp/u-boot/files/0009-board-siemens-iot2050-Generalize-the-fdt-fixup.patch:111]: Is that really a use case we want to document?

File: recipes-bsp/u-boot/files/0010-dts-iot2050-Disable-R5-lockstep-for-all-PG2-boards.patch [hunk 3]
```diff
@@ -0,0 +1,76 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Su Baocheng <baocheng.su@siemens.com>
+Date: Wed, 13 Dec 2023 10:41:09 +0800
+Subject: [PATCH] dts: iot2050: Disable R5 lockstep for all PG2 boards
+
+The R5 lockstep disabling should be common for all PG2 boards, move it
+from variants dts to common-pg2.dtsi.
+
+Signed-off-by: Su Baocheng <baocheng.su@siemens.com>
```
Comment (jan-kiszka) [inline recipes-bsp/u-boot/files/0010-dts-iot2050-Disable-R5-lockstep-for-all-PG2-boards.patch:9]: We rather need a "sync DT from kernel" commit.

File: recipes-kernel/linux/files/patches-6.1/0083-dts-iot2050-Support-IOT2050-SM-variant.patch [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
++};
++
++&usb0 {
++	pinctrl-names = "default";
++	pinctrl-0 = <&usb0_pins_default>;
++
++	maximum-speed = "high-speed";
++	/delete-property/ snps,dis-u1-entry-quirk;
++	/delete-property/ snps,dis-u2-entry-quirk;
++};
```
Comment (jan-kiszka) [inline recipes-kernel/linux/files/patches-6.1/0083-dts-iot2050-Support-IOT2050-SM-variant.patch:383]: I'm working on factoring out the Arduino bits from common.dtsi, making this file much smaller.

File: recipes-kernel/linux/files/patches-6.1/0083-dts-iot2050-Support-IOT2050-SM-variant.patch [hunk 5]
```diff
@@ -0,0 +1,383 @@
...
++	num-lanes = <1>;
++	phys = <&serdes0 PHY_TYPE_PCIE 1>;
++	phy-names = "pcie-phy0";
++	reset-gpios = <&wkup_gpio0 27 GPIO_ACTIVE_HIGH>;
++	status = "okay";
++};
++
++&pcie0_ep {
++	status = "disabled";
++};
```
Comment (jan-kiszka) [inline recipes-kernel/linux/files/patches-6.1/0083-dts-iot2050-Support-IOT2050-SM-variant.patch:359]: Already off by default.

File: recipes-kernel/linux/files/patches-6.1/0083-dts-iot2050-Support-IOT2050-SM-variant.patch [hunk 6]
```diff
@@ -0,0 +1,383 @@
...
++
++&pcie1_rc {
++	status = "disabled";
++};
++
++&pcie1_ep {
++	status = "disabled";
++};
```
Comment (jan-kiszka) [inline recipes-kernel/linux/files/patches-6.1/0083-dts-iot2050-Support-IOT2050-SM-variant.patch:367]: Also off by default.

File: recipes-kernel/linux/files/patches-6.1/0083-dts-iot2050-Support-IOT2050-SM-variant.patch [hunk 7]
```diff
@@ -0,0 +1,383 @@
...
++	pinctrl-0 = <&main_spi0_pins>;
++
++	#address-cells = <1>;
++	#size-cells= <0>;
++
++	spidev@0 {
++		compatible = "rohm,dh2228fv";
++		spi-max-frequency = <20000000>;
++		reg = <0>;
++	};
```
Comment (jan-kiszka) [inline recipes-kernel/linux/files/patches-6.1/0083-dts-iot2050-Support-IOT2050-SM-variant.patch:282]: Split the spidev node off into a separate commit - it can't be part of mainline. We still need a different solution for that problem. I'm thinking about enhancing SPI controllers by a why to instantiate a new device node during runtime via sysfs, allowing to bind spidev to that dynamically. But not code for that yet, and no idea if that will be accepted upstream.

File: recipes-kernel/linux/files/patches-6.1/0083-dts-iot2050-Support-IOT2050-SM-variant.patch [hunk 8]
```diff
@@ -0,0 +1,383 @@
...
++
++&mcu_spi0 {
++	pinctrl-names = "default";
++	pinctrl-0 = <&mcu_spi0_pins_default>;
++};
++
++&main_i2c3 {
++	pinctrl-names = "default";
++	pinctrl-0 = <&main_i2c3_pins_default>;
++	clock-frequency = <400000>;
```
Comment (jan-kiszka) [inline recipes-kernel/linux/files/patches-6.1/0083-dts-iot2050-Support-IOT2050-SM-variant.patch:293]: Redundant to what's already in common.dtsi.

File: recipes-kernel/linux/files/patches-6.1/0083-dts-iot2050-Support-IOT2050-SM-variant.patch [hunk 9]
```diff
@@ -0,0 +1,383 @@
...
++	status = "disabled";
++};
++
++&dss_ports {
++	/delete-node/ port@1;
++};
++
++&ecap0 {
++	status = "disabled";
++};
```
Comment (jan-kiszka) [inline recipes-kernel/linux/files/patches-6.1/0083-dts-iot2050-Support-IOT2050-SM-variant.patch:327]: We should add https://lore.kernel.org/linux-arm-kernel/6131d44e0505ca3efbb9039e5f2b637a3e139312.1699087938.git.jan.kiszka@siemens.com/ to the queue instead, then this is obsolete.

File: recipes-kernel/linux/files/patches-6.1/0083-dts-iot2050-Support-IOT2050-SM-variant.patch [hunk 10]
```diff
@@ -0,0 +1,383 @@
...
++&dwc3_0 {
++	assigned-clock-parents = <&k3_clks 151 4>,  /* set REF_CLK to 20MHz i.e. PER0_PLL/48 */
++				 <&k3_clks 151 9>;  /* set PIPE3_TXB_CLK to CLK_12M_RC/256 (for HS only) */
++	/delete-property/ phys;
++	/delete-property/ phy-names;
++};
++
++&usb0 {
++	pinctrl-names = "default";
++	pinctrl-0 = <&usb0_pins_default>;
```
Comment (jan-kiszka) [inline recipes-kernel/linux/files/patches-6.1/0083-dts-iot2050-Support-IOT2050-SM-variant.patch:378]: Redundant to common.dtsi

File: recipes-kernel/linux/files/patches-6.1/0078-arm64-dts-ti-iot2050-Disable-R5-lockstep-for-all-PG2.patch [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+  *   Chao Zeng <chao.zeng@siemens.com>
+@@ -9,6 +9,11 @@
+  * Common bits of the IOT2050 Basic and Advanced variants, PG2
+  */
+ 
++&mcu_r5fss0 {
++	/* lock-step mode not supported on PG2 boards */
++	ti,cluster-mode = <0>;
++};
++
```
Comment (jan-kiszka) [inline recipes-kernel/linux/files/patches-6.1/0078-arm64-dts-ti-iot2050-Disable-R5-lockstep-for-all-PG2.patch:37]: Likely not critical, but now the basic pg2 will pull this also via `k3-am6528-iot2050-basic-common.dtsi`. Maybe also move the related bits from `k3-am6528-iot2050-basic-common.dtsi` to `k3-am6528-iot2050-basic.dts`?

Comment (jan-kiszka) [reply recipes-kernel/linux/files/patches-6.1/0078-arm64-dts-ti-iot2050-Disable-R5-lockstep-for-all-PG2.patch:37]: https://github.com/siemens/linux/commit/350bbd49605c4fd26bd4c40e212cf4c800f73367

File: recipes-kernel/linux/files/patches-6.1/0083-dts-iot2050-Support-IOT2050-SM-variant.patch (hunk 4 above)
Comment (jan-kiszka) [reply recipes-kernel/linux/files/patches-6.1/0083-dts-iot2050-Support-IOT2050-SM-variant.patch:383]: https://github.com/siemens/linux/commit/136cb01bccb97d2e339e12d1a5aada551878059d
https://github.com/siemens/linux/commit/8c241baf746fac7148ac5d88da7e84f26a8d901b

File: recipes-kernel/linux/files/patches-6.1/0083-dts-iot2050-Support-IOT2050-SM-variant.patch (hunk 10 above)
Comment (BaochengSu) [reply recipes-kernel/linux/files/patches-6.1/0083-dts-iot2050-Support-IOT2050-SM-variant.patch:378]: This is a leftover after removing the usb_oc_pins... Will remove this.

File: recipes-kernel/linux/files/patches-6.1/0083-dts-iot2050-Support-IOT2050-SM-variant.patch (hunk 6 above)
Comment (BaochengSu) [reply recipes-kernel/linux/files/patches-6.1/0083-dts-iot2050-Support-IOT2050-SM-variant.patch:367]: Not default for 6.1... Let's remove it after backporting or during upsteaming

File: recipes-kernel/linux/files/patches-6.1/0083-dts-iot2050-Support-IOT2050-SM-variant.patch (hunk 5 above)
Comment (BaochengSu) [reply recipes-kernel/linux/files/patches-6.1/0083-dts-iot2050-Support-IOT2050-SM-variant.patch:359]: Not default for 6.1... Let's remove it after backporting or during upstreaming

File: recipes-kernel/linux/files/patches-6.1/0083-dts-iot2050-Support-IOT2050-SM-variant.patch (hunk 8 above)
Comment (BaochengSu) [reply recipes-kernel/linux/files/patches-6.1/0083-dts-iot2050-Support-IOT2050-SM-variant.patch:293]: Ack

File: recipes-kernel/linux/files/patches-6.1/0083-dts-iot2050-Support-IOT2050-SM-variant.patch (hunk 9 above)
Comment (BaochengSu) [reply recipes-kernel/linux/files/patches-6.1/0083-dts-iot2050-Support-IOT2050-SM-variant.patch:327]: Not default disabled for 6.1... Anyway, to make it unified and reduce the future effort, let's remove it here and  since the upstream has already disabled it by default, it should be ok for upstreaming.

Also make it okay is harmless.

File: recipes-kernel/linux/files/patches-6.1/0083-dts-iot2050-Support-IOT2050-SM-variant.patch (hunk 5 above)
Comment (BaochengSu) [reply recipes-kernel/linux/files/patches-6.1/0083-dts-iot2050-Support-IOT2050-SM-variant.patch:359]: Acked and fixed.

File: recipes-kernel/linux/files/patches-6.1/0083-dts-iot2050-Support-IOT2050-SM-variant.patch (hunk 6 above)
Comment (BaochengSu) [reply recipes-kernel/linux/files/patches-6.1/0083-dts-iot2050-Support-IOT2050-SM-variant.patch:367]: Acked and fixed.

File: recipes-bsp/u-boot/files/0008-board-siemens-iot2050-Pass-DDR-size-from-FSBL.patch (hunk 1 above)
Comment (jan-kiszka) [reply recipes-bsp/u-boot/files/0008-board-siemens-iot2050-Pass-DDR-size-from-FSBL.patch:101]: This one is still pending.

File: recipes-bsp/u-boot/files/0009-board-siemens-iot2050-Generalize-the-fdt-fixup.patch (hunk 2 above)
Comment (jan-kiszka) [reply recipes-bsp/u-boot/files/0009-board-siemens-iot2050-Generalize-the-fdt-fixup.patch:111]: And this one.

File: recipes-bsp/u-boot/files/0008-board-siemens-iot2050-Pass-DDR-size-from-FSBL.patch (hunk 1 above)
Comment (BaochengSu) [reply recipes-bsp/u-boot/files/0008-board-siemens-iot2050-Pass-DDR-size-from-FSBL.patch]: Fixed

File: recipes-bsp/u-boot/files/0009-board-siemens-iot2050-Generalize-the-fdt-fixup.patch (hunk 2 above)
Comment (BaochengSu) [reply recipes-bsp/u-boot/files/0009-board-siemens-iot2050-Generalize-the-fdt-fixup.patch]: Fixed

File: recipes-bsp/u-boot/files/0009-dts-iot2050-Sync-kernel-dts-to-u-boot.patch [hunk 12]
```diff
@@ -0,0 +1,1463 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Li Hua Qian <huaqian.li@siemens.com>
+Date: Tue, 29 Aug 2023 11:46:21 +0800
+Subject: [PATCH] dts: iot2050: Sync kernel dts to u-boot
+
```
Comment (jan-kiszka) [inline recipes-bsp/u-boot/files/0009-dts-iot2050-Sync-kernel-dts-to-u-boot.patch:5]: Would be helpful to note the source version here.

Comment (jan-kiszka) [reply recipes-bsp/u-boot/files/0009-dts-iot2050-Sync-kernel-dts-to-u-boot.patch:5]: ...which probably is 6.1 + our patches. So this patch here is downstream-only.

File: recipes-bsp/u-boot/files/0010-dts-iot2050-Support-new-IOT2050-SM-variant.patch [hunk 13]
```diff
@@ -0,0 +1,366 @@
...
++&usb0 {
++	maximum-speed = "high-speed";
++	/delete-property/ snps,dis-u1-entry-quirk;
++	/delete-property/ snps,dis-u2-entry-quirk;
++};
+diff --git a/board/siemens/iot2050/board.c b/board/siemens/iot2050/board.c
+index 32f5280125c0..8497212ab890 100644
+--- a/board/siemens/iot2050/board.c
++++ b/board/siemens/iot2050/board.c
+@@ -174,6 +174,14 @@ static bool board_is_m2(void)
```
Comment (jan-kiszka) [inline recipes-bsp/u-boot/files/0010-dts-iot2050-Support-new-IOT2050-SM-variant.patch:316]: Why is this part of the DT sync?

File: recipes-bsp/u-boot/files/0009-dts-iot2050-Sync-kernel-dts-to-u-boot.patch (hunk 12 above)
Comment (BaochengSu) [reply recipes-bsp/u-boot/files/0009-dts-iot2050-Sync-kernel-dts-to-u-boot.patch:5]: Split this patch to two? one for upstream and one for downstream?

Comment (jan-kiszka) [reply recipes-bsp/u-boot/files/0009-dts-iot2050-Sync-kernel-dts-to-u-boot.patch:5]: No, not worth to bother. We cannot predict source and target versions of upstream yet.

**Review by jan-kiszka — APPROVED** [review]
Looks good to me. I assume you tested U-Boot on SM and one or two other variants already?

**Comment by BaochengSu** [comment]
> Looks good to me. I assume you tested U-Boot on SM and one or two other variants already?

SM is tested, PG2 and co, are under testing, will let you know once finished.

**Comment by BaochengSu** [comment]
Test passed on PG1 Basic/Advanced, PG2 A/B, M2, SM.
